### DISCUSSION TITLE
Realtime on pg_dump/pg_restored schemas

### DISCUSSION THREAD

**Discussion by ltanady** [supabase/supabase#9362] [Questions]
I'm trying to get two of my pg_dump'ed schemas, cc$dev and cc$prod, working with the Supabase realtime functions. I'm able to get it working with anything in the default public schema, but I'm guessing there is something not sent in my migrated schemas that's preventing it.

Any ideas?

**Comment by kiwicopple** [comment]
Just to confirm - is Realtime (Database replication) enabled for your tables/schemas?

Some docs are here:
https://supabase.com/docs/guides/api#realtime-api

**Reply by LawJolla** [reply]
Hi @kiwicopple ,

This is a little confusing because I'm the one with the problem.  I'm a pro user and foolishly emailed @ltanady 's thinking I could get email support but that was denied because the question wasn't about "my instance."  Curious that the first question is about my instance.

Replication is enabled for the  public:test_table, $dev:Vehicle, cc$dev:Deal tables.

This feels like a problem with permissions/grants that Supabase tables automatically gain but migrated ones do not. 

Thanks for your help!

**Reply by kiwicopple** [reply]
Generally it’s good for the community if questions are solved in the open - in the same way that we build in the open, we appreciate it if the community can do the same. It’s not an obligation of course, but we get a lot of application-level questions which are less about Supabase and more about Postgres. If we only answer these questions via email then we will continue to receive emails about them. Discussions help is to solve problems for everyone because they serve as documentation for future users. 

That said, I understand the frustration. From Leo’s comment this appeared to be an application-level error with a documented solution, but if that doesn’t work then perhaps it’s a bug. I’ll check with @w3b6x9 if there is anything missing from the docs, and if not I’ll get him to debug your specific project and see if it’s a bug. Hang tight

**Reply by LawJolla** [reply]
I appreciate it!

And I understand the motivation.  But Supabase's goals are not your customer's goals.

My goal is to get a vehicle financing application done by the deadline.  I'm stressed and, I'm sorry, but it's coming through.

What I'm building has several web hooks coming in during this process and I chose Supabase realtime as a preferable solution to pinging the server with fairly heavy queries.  This is a blocking issue.

How my problem helps the community isn't really my concern.  Getting people that pay me a quality product on time is.

I will downgrade to the free tier so I don't feel taken advantage of and can stop complaining.

Thanks for your quick help, patience, and rant listening!  🤣

**Reply by w3b6x9** [reply]
@LawJolla Realtime works for all tables in the public schema as noted in the guide here: https://supabase.com/docs/guides/realtime/postgres-changes.

We have yet to provide documentation on how to listen to custom schemas because we want to dog food that internally first as well as think through the security implications before instructing our customers.

With that being said, you can get it working for your tables in custom schemas by running the following example query for the roles you want:

```sql
GRANT SELECT ON schema.table TO anon;
```

I want to mention that this only works when using `supabase-js` `v2` and Realtime has an issue at the moment when attempting to listen to tables with a space in their names.

**Reply by LawJolla** [reply]
> @LawJolla Realtime works for all tables in the public schema as noted in the guide here: https://supabase.com/docs/guides/realtime/postgres-changes.
> 
> We have yet to provide documentation on how to listen to custom schemas because we want to dog food that internally first as well as think through the security implications before instructing our customers.
> 
> With that being said, you can get it working for your tables in custom schemas by running the following example query for the roles you want:
> 
> ```sql
> GRANT SELECT ON schema.table TO anon;
> ```
> 
> I want to mention that this only works when using `supabase-js` `v2` and Realtime has an issue at the moment when attempting to listen to tables with a space in their names.

Appreciate it!

My Supabase instance isn't client exposed -- it's all traditional server calls -- so the security isn't a concern.

I ran `GRANT SELECT on "cc$dev"."Deal" TO anon` (I also tried to `anon, authenticated, service_role`) but did not get a real time notification to cc$dev.Deal table updates.

Can you confirm the following on JS V2 should have worked where the client is instantiated with the service_role key?

```
 supabase
  .channel("*")
  .on("postgres_changes", { event: "*", schema: "*" }, (payload) => {
    console.log("Change received!", payload)
  })
  .subscribe()
```

Here are my current grants for service_role and the Deal table.

![SQL Supabase 2022-10-06 at 2 47 55 PM](https://user-images.githubusercontent.com/17485128/194424784-c97316ca-f061-40be-af4b-95f6644000dc.jpg)


Thanks again!

**Reply by w3b6x9** [reply]
@LawJolla 

> it's all traditional server calls -- so the security isn't a concern.

Definitely good to know!

Your code looks good but perhaps you'll need to specify the schema and table since it's a custom schema:

```js
supabase
  .channel("*")
  .on("postgres_changes", { event: "*", schema: "cc$dev", table: "Deal" }, (payload) => {
    console.log("Change received!", payload)
  })
  .subscribe()
```

I think `*` is subscribing you to all tables in `public` schema only. If that doesn't work then we'll need to make some updates to our server and dog food some more. Appreciate you giving this a try!

**Reply by LawJolla** [reply]
I tried it both ways.  Please see the attached video.  (Hopefully I didn't leak any secrets 😬)


https://user-images.githubusercontent.com/17485128/194427117-e8ce108e-d149-4c8d-8b3f-27eefffcaae0.mp4

**Reply by w3b6x9** [reply]
@LawJolla thanks for sharing the video!

Looks like it works for tables in the `public` schema only. I'll investigate and get back to you.

**Reply by LawJolla** [reply]
Thank you @w3b6x9 

@ltanady it sounds like we have "specific issues relating to my instance."

**Reply by w3b6x9** [reply]
I just tried this and it works when both `schema` is set to `*` and specifically setting the `schema` and `table`:

```js
 supabase
  .channel("room")
  .on("postgres_changes", { event: "*", schema: "*" }, (payload) => {
    console.log("Change received on *:", payload)
  })
  .on("postgres_changes", { event: "*", schema: "custom", table: "test" }, (payload) => {
    console.log("Change received on custom schema:", payload)
  })
  .subscribe()
```

**Comment by LawJolla** [comment]
Thanks @w3b6x9 for your quick follow up.

I'm not sure what you mean when you say it works.  What is working?


Here's a demo of me trying it again.  Please let me know what I'm missing.


https://user-images.githubusercontent.com/17485128/194473415-27112572-dace-4aa9-99d5-96495ea1bcca.mp4

**Reply by LawJolla** [reply]
Trying a few different things, I tried the following.

``create schema custom;``

Used the Supabase UI to create a table `test_table` with id, created_at, and test fields.

Turned on table replication.

Ran ``supabase
  .channel("*")
  .on("postgres_changes", { event: "*", schema: "*" }, (payload) => {
    console.log("Change received, sub 2!", payload)
  })
  .subscribe()``

and got 401 / unauthed response. 

``GRANT SELECT ON custom.test TO service_role;``

And everything works.

My migrated custom schemas don't get any 401 response.

What is different about those from how Supabase configures green field custom schemas?

**Comment by LawJolla** [comment]
The problem is there's a Supabase bug parsing the schema's dollar sign.

Uploading Screen Recording 2022-10-07 at 12.57.50 PM.mp4…

**Reply by w3b6x9** [reply]
@LawJolla are you receiving changes now that you've removed the dollar sign from schema name?

I just created a new schema with a dollar sign in its name and I'm able to get changes.

**Reply by LawJolla** [reply]
Yes.  And when I add back the dollar sign, realtime stops.  It doesn't look like my screen recording attached.  Let me try it again.

https://user-images.githubusercontent.com/17485128/194648103-3b6a857e-5ac7-41f0-90b8-af94f9e84964.mp4

**Comment by w3b6x9** [comment]
@LawJolla Hey, just wanted to follow up and ask how you created the schema `cc$dev` and the tables under that schema prior to the dump. Did you run a couple SQL commands or use something like Prisma?

**Reply by LawJolla** [reply]
cc$dev is just a clone of a recent cc$prod (I was too lazy to make mock data in the beginning and wow do I regret that decision now 🤣)

cc$prod is an old Prisma 1 schema, circa 2018, that's been migrated to Prisma 3.

**Comment by w3b6x9** [comment]
@LawJolla I think the/another issue you're facing is that Realtime currently doesn't work with columns that have types with uppercase letters in them. We just became aware of this issue: https://github.com/supabase/realtime/issues/306.

The Realtime team is aware and we're going to address it as soon as we can.

**Reply by LawJolla** [reply]
Thanks @w3b6x9 

I haven't pushed the realtime into production (or my production schema, `cc$prod`, but on my dev schema `cc_dev` the realtime works with upper case with my code above.

**Reply by w3b6x9** [reply]
We just deployed a fix that addresses https://github.com/supabase/realtime/issues/306.
